onnxruntime/test/framework/inference_session_test.cc

```diff
@@ -1224,6 +1244,88 @@ TEST(ExecutionProviderTest, FunctionTest) {
+
+  auto& input_arg_1 = graph.GetOrCreateNodeArg("X", &float_tensor);
+  auto& input_arg_2 = graph.GetOrCreateNodeArg("Y", &float_tensor);
+  inputs.push_back(&input_arg_1);
+  inputs.push_back(&input_arg_2);
+  auto& output_arg = graph.GetOrCreateNodeArg("node_1_out_1", &float_tensor);
+  outputs.push_back(&output_arg);
+  graph.AddNode("node_1", "Add", "node 1.", inputs, outputs);
+
+  auto& input_arg_3 = graph.GetOrCreateNodeArg("Z", &float_tensor);
+  inputs.clear();
+  inputs.push_back(&output_arg);
+  inputs.push_back(&input_arg_3);
+  auto& output_arg_2 = graph.GetOrCreateNodeArg("M", &float_tensor);
+  outputs.clear();
+  outputs.push_back(&output_arg_2);
+  graph.AddNode("node_2", "Add", "node 2.", inputs, outputs);
+
+  auto status = graph.Resolve();
+  ASSERT_TRUE(status.IsOK());
+  std::string model_file_name = "execution_provider_test_graph.onnx";
+  status = onnxruntime::Model::Save(model, model_file_name);
+
+  SessionOptions so;
+  so.session_logid = "ExecutionProviderTest.FunctionTest";
+  InferenceSession session_object{so, GetEnvironment()};
+  status = session_object.Load(model_file_name);
+  ASSERT_TRUE(status.IsOK());
+  status = session_object.Initialize();
+  ASSERT_TRUE(status.IsOK());
+
+  RunOptions run_options;
+  run_options.run_tag = so.session_logid;
+
+  CPUExecutionProviderInfo epi;
+  auto testCPUExecutionProvider = std::make_unique<::onnxruntime::CPUExecutionProvider>(epi);
+
+  std::vector<int64_t> dims_mul_x = {3, 2};
+  std::vector<float> values_mul_x = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};
+  OrtValue ml_value_x;
+  CreateMLValue<float>(testCPUExecutionProvider->GetAllocator(0, OrtMemTypeDefault), dims_mul_x, values_mul_x, &ml_value_x);
+  OrtValue ml_value_y;
+  CreateMLValue<float>(testCPUExecutionProvider->GetAllocator(0, OrtMemTypeDefault), dims_mul_x, values_mul_x, &ml_value_y);
+  OrtValue ml_value_z;
+  CreateMLValue<float>(testCPUExecutionProvider->GetAllocator(0, OrtMemTypeDefault), dims_mul_x, values_mul_x, &ml_value_z);
+  NameMLValMap feeds;
+  feeds.insert(std::make_pair("X", ml_value_x));
+  feeds.insert(std::make_pair("Y", ml_value_y));
+  feeds.insert(std::make_pair("Z", ml_value_z));
+
+  // prepare outputs
+  std::vector<std::string> output_names;
+  output_names.push_back("M");
+  std::vector<OrtValue> fetches;
+
+  TestingInferenceSession session{so, GetEnvironment()};
+  ASSERT_STATUS_OK(
+      session.RegisterExecutionProvider(std::make_unique<::onnxruntime::FuseExecutionProvider>()));
+  status = session.Load(model_file_name);
+  ASSERT_TRUE(status.IsOK());
+  status = session.Initialize();
+  ASSERT_TRUE(status.IsOK());
+
+  const Graph& fused_graph = session.GetModel().MainGraph();
+  ASSERT_TRUE(fused_graph.NumberOfNodes() == 1);
+  for (const Node& node : fused_graph.Nodes()) {
+    ASSERT_TRUE(node.NodeType() == Node::Type::Fused);
+    ASSERT_TRUE(node.Op()->has_type_and_shape_inference_function());
+  }
```

Comment:
I would have expected the test to check the inferred output shape after Graph::Resolve is called. actually executing the model seems optional, and doesn't necessarily prove the type/shape inferencing worked as we handle unknown output shapes. 

should be able to look at the output definitions for the fused node to check the shape is set and is as expected (using Node::OutputDefs() and NodeArg::Shape()).

Reply:
Done. The test now loads the graph with the FusedExecutionProvider enabled, calls `.Resolve()` to perform a fresh round of shape-inference, then verifies that the shape is as expected.